## What is Feature Engineering?

#### Feature Engineering is the process of creating new, meaningful columns from existing data to make analysis, reporting, SQL queries, dashboards, and AI insights more useful.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("../data/cleaned/employee_cleaned.csv")

df.head()



,age,attrition_status,business_travel_frequency,daily_rate,department,distance_from_home,education_level,education_field,employee_id,environment_satisfaction,...,performance_rating,relationship_satisfaction,stock_option_level,total_working_years,training_sessions_last_year,work_life_balance,years_at_company,years_in_current_role,years_since_last_promotion,years_with_current_manager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,2,...,3,1,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,2,3,...,4,4,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,4,4,...,3,2,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,5,4,...,3,3,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,7,1,...,3,4,1,6,3,3,2,2,2,2


In [2]:

choices = [
    "Young Professional",
    "Early Career",
    "Mid Career",
    "Senior Professional"
]

In [3]:

conditions = [
    (df["age"] >= 18) & (df["age"] <= 25),
    (df["age"] >= 26) & (df["age"] <= 35),
    (df["age"] >= 36) & (df["age"] <= 45),
    (df["age"] >= 46) & (df["age"] <= 60)
]


In [4]:
choices = [
    "Young Professional",
    "Early Career",
    "Mid Career",
    "Senior Professional"
]

In [5]:
df["age_group"] = np.select(
    conditions,
    choices,
    default="Unknown"
)

### Verify the Result


In [6]:
df[["age", "age_group"]].head(10)

,age,age_group
0,41,Mid Career
1,49,Senior Professional
2,37,Mid Career
3,33,Early Career
4,27,Early Career
5,32,Early Career
6,59,Senior Professional
7,30,Early Career
8,38,Mid Career
9,36,Mid Career


### Count Employees in Each Group

In [7]:
df["age_group"].value_counts()

age_group
Early Career           606
Mid Career             468
Senior Professional    273
Young Professional     123
Name: count, dtype: int64

### Validate

In [8]:
df[df["age_group"] == "Unknown"]

,age,attrition_status,business_travel_frequency,daily_rate,department,distance_from_home,education_level,education_field,employee_id,environment_satisfaction,...,relationship_satisfaction,stock_option_level,total_working_years,training_sessions_last_year,work_life_balance,years_at_company,years_in_current_role,years_since_last_promotion,years_with_current_manager,age_group


In [9]:
df["monthly_income"].describe()

count     1470.000000
mean      6502.931293
std       4707.956783
min       1009.000000
25%       2911.000000
50%       4919.000000
75%       8379.000000
max      19999.000000
Name: monthly_income, dtype: float64

In [10]:
salary_conditions = [
    (df["monthly_income"] <= 2911),
    (df["monthly_income"] > 2911) & (df["monthly_income"] <= 4919),
    (df["monthly_income"] > 4919) & (df["monthly_income"] <= 8379),
    (df["monthly_income"] > 8379)
]

### Labels

In [11]:
salary_choices = [
    "Low Income",
    "Medium Income",
    "High Income",
    "Very High Income"
]

### Create the Feature

In [12]:
df["salary_band"] = np.select(
    salary_conditions,
    salary_choices,
    default="Unknown"
)

### Verify

In [13]:
df[["monthly_income", "salary_band"]].head(10)

,monthly_income,salary_band
0,5993,High Income
1,5130,High Income
2,2090,Low Income
3,2909,Low Income
4,3468,Medium Income
5,3068,Medium Income
6,2670,Low Income
7,2693,Low Income
8,9526,Very High Income
9,5237,High Income


In [14]:
df["salary_band"].value_counts()

salary_band
Low Income          369
Very High Income    368
High Income         367
Medium Income       366
Name: count, dtype: int64

### Why did you use these salary ranges?"

##### You can answer:

##### "Instead of using arbitrary salary thresholds, I analyzed the salary distribution using quartiles. I created salary bands based on the 25th, 50th, and 75th percentiles, ensuring that the categories reflected the actual distribution of employee salaries in the dataset. This approach produces balanced groups and more meaningful business analysis."

🚀 Feature 3 – Experience Level
🏢 Business Problem

The HR Manager asks:

"Are junior employees leaving the company more frequently than senior employees?"

| Years | Experience Level | Reason               |
| ----: | ---------------- | -------------------- |
|   0–5 | Junior           | New to the workforce |
|  6–10 | Mid-Level        | Developing expertise |
| 11–20 | Senior           | Highly experienced   |
|   21+ | Expert           | Veteran employees    |


In [15]:
df["total_working_years"].describe()

count    1470.000000
mean       11.279592
std         7.780782
min         0.000000
25%         6.000000
50%        10.000000
75%        15.000000
max        40.000000
Name: total_working_years, dtype: float64

Conditions

In [16]:
experience_conditions = [
    (df["total_working_years"] <= 5),
    (df["total_working_years"] > 5) & (df["total_working_years"] <= 10),
    (df["total_working_years"] > 10) & (df["total_working_years"] <= 20),
    (df["total_working_years"] > 20)
]

Labels

In [17]:
experience_choices = [
    "Junior",
    "Mid-Level",
    "Senior",
    "Expert"
]

Create the Feature

In [18]:
df["experience_level"] = np.select(
    experience_conditions,
    experience_choices,
    default="Unknown"
)

Verify

In [19]:
df[["total_working_years", "experience_level"]].head(10)

,total_working_years,experience_level
0,8,Mid-Level
1,10,Mid-Level
2,7,Mid-Level
3,8,Mid-Level
4,6,Mid-Level
5,8,Mid-Level
6,12,Senior
7,1,Junior
8,10,Mid-Level
9,17,Senior


Count Employees

In [20]:
df["experience_level"].value_counts()

experience_level
Mid-Level    607
Senior       340
Junior       316
Expert       207
Name: count, dtype: int64

🎯 Interview Answer

Interviewer: Why did you create experience_level?

Answer:

"The dataset contained raw years of work experience, but HR analysis is more meaningful when employees are grouped into experience levels. I engineered an experience_level feature to segment employees into Junior, Mid-Level, Senior, and Expert categories. This enables easier analysis of attrition, salary, promotions, and performance across different workforce segments."

🚀 Next Feature (Feature 4)

I suggest we build:

⭐ promotion_due
Business Problem

The HR Manager asks:

"Which employees should be considered for promotion?"

We'll create a business rule using:

years_since_last_promotion
performance_rating
job_level
| Column                       | Why?                               |
| ---------------------------- | ---------------------------------- |
| `years_since_last_promotion` | How long since the last promotion? |
| `performance_rating`         | Is the employee performing well?   |
| `job_level`                  | Current level in the organization. |

Business Logic

Before writing code, we define the rule.

For example:

| Years Since Last Promotion | Performance Rating | Promotion Due |
| -------------------------: | -----------------: | ------------- |
|                          6 |                  4 | Yes           |
|                          7 |                  3 | Yes           |
|                          2 |                  4 | No            |
|                          5 |                  2 | No            |

If:

years_since_last_promotion >= 5
AND
performance_rating >= 3

In [21]:
df["years_since_last_promotion"].describe()
df["performance_rating"].describe()


count    1470.000000
mean        3.153741
std         0.360824
min         3.000000
25%         3.000000
50%         3.000000
75%         3.000000
max         4.000000
Name: performance_rating, dtype: float64

In [22]:
# Create Promotion Due Feature

df["promotion_due"] = np.where(
    (df["years_since_last_promotion"] >= 5) &
    (df["performance_rating"] >= 3),
    "Yes",
    "No"
)

In [23]:
df[[
    "years_since_last_promotion",
    "performance_rating",
    "promotion_due"
]].head(20)

,years_since_last_promotion,performance_rating,promotion_due
0,0,3,No
1,1,4,No
2,0,3,No
3,3,3,No
4,2,3,No
5,3,3,No
6,0,4,No
7,0,4,No
8,1,4,No
9,7,3,Yes


### Count Employees

In [24]:
df["promotion_due"].value_counts()

promotion_due
No     1210
Yes     260
Name: count, dtype: int64

Validation

In [25]:
df["promotion_due"].unique()

<ArrowStringArray>
['No', 'Yes']
Length: 2, dtype: str

🎤 Interview Answer

"I engineered a promotion_due feature using NumPy's np.where() function. The feature identifies employees who may be eligible for promotion based on a business rule combining years since the last promotion and performance rating. This enables HR to analyze promotion readiness and supports workforce planning."

🧠 Why did we use np.where() instead of np.select()?

Because this is a binary decision.

Two outcomes:
✅ Yes
✅ No

For binary conditions, np.where() is simpler and more readable.

Situation	NumPy Function
2 outcomes (Yes/No)	np.where() ✅
3 or more outcomes	np.select() ✅



# 🚀 Feature 5 – Income Per Year of Experience

This is one of my favorite features because **it doesn't exist in the original dataset**, yet it provides valuable business insight.

## 🏢 Business Problem

Imagine the HR Director asks:

> **"Are employees with similar experience being paid fairly?"**

Suppose we have two employees:

| Employee | Monthly Income | Total Working Years |
| -------- | -------------: | ------------------: |
| A        |         12,000 |                  12 |
| B        |         12,000 |                   3 |

Both earn ₹12,000.

But Employee B earns much more **relative to experience**.

---

# 🎯 Business Logic

We create a new metric:

```text
Income Per Year of Experience
=
Monthly Income
────────────────────────────
Total Working Years
```

Example:

| Income | Experience | Result |
| -----: | ---------: | -----: |
|  12000 |         12 |   1000 |
|  12000 |          3 |   4000 |

Employee B has a much higher income per year of experience.

---

# ⚠️ Important

Some employees have:

```text
total_working_years = 0
```

If we divide by 0:

```python
12000 / 0
```

Python throws an error (or produces infinity).

So we need to handle that safely.








Example:

| monthly_income | total_working_years | income_per_year_experience |
| -------------: | ------------------: | -------------------------: |
|           5993 |                   8 |                     749.12 |
|           5130 |                  10 |                     513.00 |
|           2090 |                   7 |                     298.57 |



# 🎤 Interview Answer

> "I engineered an `income_per_year_experience` feature to measure salary relative to experience. This provides a normalized metric that can help identify employees who may be overpaid or underpaid compared to peers with similar experience. I used `np.where()` to safely handle employees with zero years of experience and avoid division-by-zero errors."

---

# ⭐ Why this feature is impressive

This isn't just another category like `age_group`.

It's a **derived business metric**, which demonstrates analytical thinking. It can support:

* Compensation analysis
* Salary fairness studies
* HR decision-making
* AI-generated recommendations

---

# 📊 Current Progress

That's how experienced analysts think. Quality matters more than quantity. I expect we'll finish Phase 6 with **7–8 strong features**, not 15 average ones. That will make your project much stronger.


# 💻 Step 1 – Create the Feature

We'll use **`np.where()`** to avoid division by zero.


In [26]:

df["income_per_year_experience"] = np.where(
    df["total_working_years"] == 0,
    df["monthly_income"],
    df["monthly_income"] / df["total_working_years"]
)

# 💻 Step 2 – Round the Values


In [27]:
df["income_per_year_experience"] = df["income_per_year_experience"].round(2)

# 💻 Step 3 – Verify


In [28]:

df[[
    "monthly_income",
    "total_working_years",
    "income_per_year_experience"
]].head(10)

,monthly_income,total_working_years,income_per_year_experience
0,5993,8,749.12
1,5130,10,513.00
2,2090,7,298.57
3,2909,8,363.62
4,3468,6,578.00
5,3068,8,383.50
6,2670,12,222.50
7,2693,1,2693.00
8,9526,10,952.60
9,5237,17,308.06


# 💻 Step 4 – Check Statistics

In [29]:
df["income_per_year_experience"].describe()

count    1470.000000
mean      718.266367
std       485.321710
min       100.050000
25%       417.800000
50%       602.815000
75%       849.945000
max      2994.000000
Name: income_per_year_experience, dtype: float64

Awesome! 🔥 We now have **5 professional engineered features**.

Let's continue with a feature that HR departments actually use.

---

# 🚀 Feature 6 – Employee Tenure Category

## 🏢 Business Problem

The HR Manager asks:

> **"How many new employees, experienced employees, and long-term employees do we have?"**

The dataset has:

```text
years_at_company
```

Example:

| years_at_company |
| ---------------: |
|                1 |
|                5 |
|               12 |
|               25 |

These are numbers.

For business reporting, it's easier to analyze categories.

---

# 🎯 Business Rule

| Years at Company | Tenure Category      |
| ---------------: | -------------------- |
|              0–2 | New Employee         |
|              3–5 | Established Employee |
|             6–10 | Experienced Employee |
|              11+ | Veteran Employee     |

This helps HR answer questions like:

* Which tenure group has the highest attrition?
* Are new employees leaving too early?
* Do veteran employees receive more promotions?

# 🎤 Interview Answer

> "I created an `employee_tenure_category` feature to segment employees based on their tenure with the organization. This helps HR analyze retention, promotion trends, and attrition across different tenure groups instead of relying only on raw numerical values."




In [30]:


# 💻 Step 1 – Create Conditions

tenure_conditions = [
    (df["years_at_company"] <= 2),
    (df["years_at_company"] >= 3) & (df["years_at_company"] <= 5),
    (df["years_at_company"] >= 6) & (df["years_at_company"] <= 10),
    (df["years_at_company"] >= 11)
]


# 💻 Step 2 – Create Labels


tenure_choices = [
    "New Employee",
    "Established Employee",
    "Experienced Employee",
    "Veteran Employee"
]


# 💻 Step 3 – Create the Feature

df["employee_tenure_category"] = np.select(
    tenure_conditions,
    tenure_choices,
    default="Unknown"
)


# 💻 Step 4 – Verify

df[[
    "years_at_company",
    "employee_tenure_category"
]].head(10)




# 💻 Step 5 – Count

df["employee_tenure_category"].value_counts()



employee_tenure_category
Experienced Employee    448
Established Employee    434
New Employee            342
Veteran Employee        246
Name: count, dtype: int64

In [31]:
conditions = [
    df["monthly_income"] < 5000,
    (df["monthly_income"] >= 5000) & (df["monthly_income"] < 10000),
    df["monthly_income"] >= 10000
]

choices = ["Low Income", "Medium Income", "High Income"]

df["salary_band"] = np.select(conditions, choices, default="Unknown")


# ⭐ Feature 7 – Attrition Risk Score

This isn't machine learning yet. It's a **rule-based AI feature**, similar to what many companies build before training predictive models.

---

# 🏢 Business Problem

Imagine the HR Director asks:

> **"Can we identify employees who are at a high risk of leaving before they resign?"**

The original dataset doesn't have this answer.

So we'll create it.

---

# 🧠 Which columns should we use?

We should only use columns that have a logical relationship with attrition.

| Column              | Why it matters                                          |
| ------------------- | ------------------------------------------------------- |
| `overtime_status`   | Employees working overtime may be more likely to leave. |
| `job_satisfaction`  | Low satisfaction can increase attrition risk.           |
| `work_life_balance` | Poor balance often contributes to turnover.             |
| `promotion_due`     | Waiting too long for promotion can reduce engagement.   |

Notice I **didn't** include age or gender because there's no general business rule that those alone determine attrition risk.

---

# 🎯 Business Rules

### 🔴 High Risk

Employee is high risk if:

* Overtime = Yes
* Job Satisfaction ≤ 2
* Work-Life Balance ≤ 2
* Promotion Due = Yes

---

### 🟡 Medium Risk

Employee meets **any two** of the above conditions.

---

### 🟢 Low Risk

Everyone else.

---



---

# 🎤 Interview Answer

If the interviewer asks:

> **"How did you create the Attrition Risk Score?"**

You can answer:

> "I designed a rule-based attrition risk score by combining four HR factors: overtime status, job satisfaction, work-life balance, and promotion eligibility. Employees meeting all four risk conditions were classified as High Risk, those meeting at least two conditions as Medium Risk, and the remaining employees as Low Risk. This engineered feature was later used to generate AI-driven workforce insights."

---

# ⭐ Why this feature is special

This is no longer simple feature engineering.

You're building a **Decision Support System**.

```
Employee Data
       │
       ▼
Business Rules
       │
       ▼
Attrition Risk Score
       │
       ▼
AI Insights
       │
       ▼
Executive Dashboard


In [32]:
# 💻 Step 1 – Conditions

risk_conditions = [
    (
        (df["over_time"] == "Yes") &
        (df["job_satisfaction"] <= 2) &
        (df["work_life_balance"] <= 2) &
        (df["promotion_due"] == "Yes")
    ),

    (
        (
            (df["over_time"] == "Yes").astype(int) +
            (df["job_satisfaction"] <= 2).astype(int) +
            (df["work_life_balance"] <= 2).astype(int) +
            (df["promotion_due"] == "Yes").astype(int)
        ) >= 2
    )
]




In [33]:
# 💻 Step 2 – Labels


risk_choices = [
    "High Risk",
    "Medium Risk"
]


# 💻 Step 3 – Create the Feature


df["attrition_risk_score"] = np.select(
    risk_conditions,
    risk_choices,
    default="Low Risk"
)




# 💻 Step 4 – Verify

df[[
    "over_time",
    "job_satisfaction",
    "work_life_balance",
    "promotion_due",
    "attrition_risk_score"
]].head(15)


# 💻 Step 5 – Count

df["attrition_risk_score"].value_counts()


attrition_risk_score
Low Risk       1005
Medium Risk     455
High Risk        10
Name: count, dtype: int64

In [34]:
print("Shape:", df.shape)
print()

print("Total Columns:", len(df.columns))
print()

print(df.columns.tolist())

Shape: (1470, 39)

Total Columns: 39

['age', 'attrition_status', 'business_travel_frequency', 'daily_rate', 'department', 'distance_from_home', 'education_level', 'education_field', 'employee_id', 'environment_satisfaction', 'gender', 'hourly_rate', 'job_involvement_level', 'job_level', 'job_role', 'job_satisfaction', 'marital_status', 'monthly_income', 'monthly_rate', 'companies_worked_count', 'over_time', 'salary_hike_percent', 'performance_rating', 'relationship_satisfaction', 'stock_option_level', 'total_working_years', 'training_sessions_last_year', 'work_life_balance', 'years_at_company', 'years_in_current_role', 'years_since_last_promotion', 'years_with_current_manager', 'age_group', 'salary_band', 'experience_level', 'promotion_due', 'income_per_year_experience', 'employee_tenure_category', 'attrition_risk_score']


In [35]:
print(df.columns.tolist())

['age', 'attrition_status', 'business_travel_frequency', 'daily_rate', 'department', 'distance_from_home', 'education_level', 'education_field', 'employee_id', 'environment_satisfaction', 'gender', 'hourly_rate', 'job_involvement_level', 'job_level', 'job_role', 'job_satisfaction', 'marital_status', 'monthly_income', 'monthly_rate', 'companies_worked_count', 'over_time', 'salary_hike_percent', 'performance_rating', 'relationship_satisfaction', 'stock_option_level', 'total_working_years', 'training_sessions_last_year', 'work_life_balance', 'years_at_company', 'years_in_current_role', 'years_since_last_promotion', 'years_with_current_manager', 'age_group', 'salary_band', 'experience_level', 'promotion_due', 'income_per_year_experience', 'employee_tenure_category', 'attrition_risk_score']


In [36]:
df.to_csv("../data/featured/employee_featured.csv", index=False)

In [37]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 39 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   age                          1470 non-null   int64  
 1   attrition_status             1470 non-null   str    
 2   business_travel_frequency    1470 non-null   str    
 3   daily_rate                   1470 non-null   int64  
 4   department                   1470 non-null   str    
 5   distance_from_home           1470 non-null   int64  
 6   education_level              1470 non-null   int64  
 7   education_field              1470 non-null   str    
 8   employee_id                  1470 non-null   int64  
 9   environment_satisfaction     1470 non-null   int64  
 10  gender                       1470 non-null   str    
 11  hourly_rate                  1470 non-null   int64  
 12  job_involvement_level        1470 non-null   int64  
 13  job_level                    

In [38]:
print(df.dtypes.to_string())

age                              int64
attrition_status                   str
business_travel_frequency          str
daily_rate                       int64
department                         str
distance_from_home               int64
education_level                  int64
education_field                    str
employee_id                      int64
environment_satisfaction         int64
gender                             str
hourly_rate                      int64
job_involvement_level            int64
job_level                        int64
job_role                           str
job_satisfaction                 int64
marital_status                     str
monthly_income                   int64
monthly_rate                     int64
companies_worked_count           int64
over_time                          str
salary_hike_percent              int64
performance_rating               int64
relationship_satisfaction        int64
stock_option_level               int64
total_working_years      

✅ Connected Successfully!
